gst/tensor_ros_src/tensor_ros_src.cc

```diff
@@ -0,0 +1,464 @@
+      rossrc->count = get_item_count (rossrc->input_dims);
+      GST_DEBUG_OBJECT (rossrc, "Item count: %u\n", rossrc->count);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+/**
+ * @brief Getter for tensor_ros_src properties
+ */
+static void
+gst_tensor_ros_src_get_property (GObject * object, guint prop_id,
+    GValue * value, GParamSpec * pspec)
+{
+  GstTensorRosSrc *rossrc = GST_TENSOR_ROS_SRC (object);
+
+  switch (prop_id) {
+    case PROP_SILENT:
+      g_value_set_boolean (value, rossrc->silent);
+      break;
+    
+    case PROP_TOPIC:
+      g_value_set_string (value, rossrc->topic_name);
+      break;
+
+    case PROP_FREQ_RATE:
+      g_value_set_ulong (value, G_USEC_PER_SEC / rossrc->freq_rate);
+      break;
+
+    case PROP_DATATYPE:
+      g_value_set_string (value, get_string_type(rossrc->datatype));
+      break;
+
+    case PROP_INPUT_DIMS:
+      g_value_set_string (value, rossrc->input_dims);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+      break;
+  }
+}
+
+/**
+ * @brief Getter for tensor_ros_src properties
+ */
+static GstStateChangeReturn
+gst_tensor_ros_src_change_state (GstElement * element, GstStateChange transition)
+{
+  GstTensorRosSrc *rossrc = GST_TENSOR_ROS_SRC (element);
+  GstStateChangeReturn ret;
+
+  /* handle before change */
+  switch (transition) {
+    case GST_STATE_CHANGE_NULL_TO_READY:
+      /* create thread */
+      GST_DEBUG_OBJECT (rossrc, "State is changed: GST_STATE_CHANGE_NULL_TO_READY\n");
+      rossrc->ros_sub = new TensorRosSub ("TensorRosSub", rossrc->topic_name,
+        rossrc, rossrc->freq_rate);
+      rossrc->ros_sub->Start (&rossrc->thread);
+      break;
+
+    default:
+      break;
+  }
+
+  ret = GST_ELEMENT_CLASS (parent_class)->change_state (element, transition);
+
+  /* handle after change */
+  switch (transition) {
+    case GST_STATE_CHANGE_READY_TO_NULL:
+      /* stop thread */
+      GST_DEBUG_OBJECT (rossrc, "State is changed: GST_STATE_CHANGE_READY_TO_NULL\n");
+      rossrc->ros_sub->RequestStop ();
+      break;
+
+    default:
+      break;
+  }
+
+  return ret;
+}
+
+/**
+ * @brief Push GstBuffer which contains the subscribed ROS data
+ */
+static GstFlowReturn
+gst_tensor_ros_src_create (GstPushSrc * src, GstBuffer ** buffer)
+{
+  gpointer queue_item = NULL;
+  GstTensorRosSrc *rossrc = GST_TENSOR_ROS_SRC (src);
+  GstBuffer *buf = NULL;
+  GstMemory *mem;
+  GstMapInfo info;
+  gsize size = rossrc->count * tensor_element_size[rossrc->datatype];
+
+  /* get item from queue */
+  queue_item = g_async_queue_try_pop (rossrc->queue);
+
+  buf = gst_buffer_new ();
+  mem = gst_allocator_alloc (NULL, size, NULL);
+  gst_memory_map (mem, &info, GST_MAP_WRITE);
+
+  if (queue_item == NULL) {
+    memset (info.data, 0x00, size);
+  } else {
+    std::memcpy(info.data, queue_item, size);
```

Comment:
IMHO, this memcpy could be eliminated by directly using queue_item instead of mem.